### PR TITLE
add the pod delete feature to the kubernetes react plugin

### PR DESCRIPTION
**Pull request by sebalaini**
## Hey, I just made a Pull Request!

This PR is for the issue: https://github.com/backstage/backstage/issues/24293 

- add the pod delete feature to the kubernetes react plugin.

these were spotted in the terminal while running the test suite:

- I fixed an "error" in the `FixDialog.test`, the `data-testid` is not allowed in a fragment.
- I fixed an "error" in the `Events.tsx`, the `ListItem ` key in the map was not unique.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### Default button text
image

### Custom text in the button
image

### Delete dialog
image

### CODE REVIEW

File: plugins/kubernetes-react/src/components/Pods/Events/Events.tsx
Comment (sebalaini) [inline plugins/kubernetes-react/src/components/Pods/Events/Events.tsx:88]: the `ListItem` key was flagged as not unique.

File: plugins/kubernetes-react/src/components/Pods/FixDialog/FixDialog.test.tsx
Comment (sebalaini) [inline plugins/kubernetes-react/src/components/Pods/FixDialog/FixDialog.test.tsx:24]: the `data-testid` was flagged as not allowed in a fragment during the tests.

**Comment by mclarke47** [comment]
Hey thanks for the PR! Since this is a pod level command can we put it at the top of the drawer rather than inside the container panels? Like so?

Screenshot 2024-04-29 at 2 05 14 PM

**Comment by LegendSebastianoL** [comment]
> Hey thanks for the PR! Since this is a pod level command can we put it at the top of the drawer rather than inside the container panels? Like so?
> 
> Screenshot 2024-04-29 at 2 05 14 PM

hey :) sure I can move it to the top. I'll do it tonight

**Comment by sebalaini** [comment]
the button has been moved :)

image

**Review by mclarke47 — CHANGES_REQUESTED** [review]
looking good so far, 1 suggestion for the config and some unnecessary react components.

File: plugins/kubernetes-react/src/components/Pods/PodDelete/PodDeleteDialog.tsx
Comment (sebalaini) [reply plugins/kubernetes-react/src/components/Pods/PodDelete/PodDeleteDialog.tsx:54]: I think I understand you, the only question is once you push the button what would be the expected behavior? show like a success/error banner on the entity page? in the PodDrawer? shall the PodDrawer close once the button is clicked and then show a banner or something?

File: docs/features/kubernetes/configuration.md
Comment (sebalaini) [reply docs/features/kubernetes/configuration.md:20]: do you mean wrapping this and the podExecTerminal like this?
```
kubernetes?: {
    frontend: {
       podExecTerminal?: {
          enabled?: boolean;
       };
       podDelete?: {
       ...
   }
```

File: plugins/kubernetes-react/src/components/Pods/PodDelete/PodDeleteDialog.tsx
Comment (mclarke47) [reply plugins/kubernetes-react/src/components/Pods/PodDelete/PodDeleteDialog.tsx:54]: I think think we can change the delete icon to a `<CircularProgress />` then if the delete succeeds we close the drawer (which might happen automatically, might need to test this)

Comment (sebalaini) [reply plugins/kubernetes-react/src/components/Pods/PodDelete/PodDeleteDialog.tsx:54]: OK let me try

Comment (sebalaini) [reply plugins/kubernetes-react/src/components/Pods/PodDelete/PodDeleteDialog.tsx]: I pushed a new change, let me know how it looks.

File: plugins/kubernetes-react/src/components/Pods/PodDelete/PodDeleteButton.tsx
Comment (sebalaini) [inline plugins/kubernetes-react/src/components/Pods/PodDelete/PodDeleteButton.tsx:48]: this is deliberately set to true and never reverted because at the end of the await the drawer closes so there is no need to set it back

**Comment by sebalaini** [comment]
hey @mclarke47 just a heads up that I can make changes till Tuesday morning, then I'll not be available till the end of the month, around the 28th.

**Comment by mclarke47** [comment]
Nice! Just the config change and I think we're ready to ship

File: docs/features/kubernetes/configuration.md
Comment (sebalaini) [reply docs/features/kubernetes/configuration.md]: OK :) this is now done, all pushed.

File: plugins/kubernetes-react/src/components/Pods/PodDelete/PodDeleteButton.tsx
Comment (mclarke47) [inline plugins/kubernetes-react/src/components/Pods/PodDelete/PodDeleteButton.tsx:53]: Can we bubble this error to the UI somewhere?

Comment (sebalaini) [reply plugins/kubernetes-react/src/components/Pods/PodDelete/PodDeleteButton.tsx:53]: I guess so, I could add an extra line below the button in case of an error, and say to check the console for more info.

Comment (sebalaini) [reply plugins/kubernetes-react/src/components/Pods/PodDelete/PodDeleteButton.tsx:59]: that's done :)

Comment (sebalaini) [reply plugins/kubernetes-react/src/components/Pods/PodDelete/PodDeleteButton.tsx:59]: this is how it look like.
 
Screenshot 2024-05-03 at 19 24 46

**Review by mclarke47 — APPROVED** [review]
Thanks so much for the contribution!

**Comment by mclarke47** [comment]
@Rugvip got a sec?

File: plugins/kubernetes-react/config.d.ts
Comment (Rugvip) [inline plugins/kubernetes-react/config.d.ts:47]: Could you use the i18n APIs for this instead? It covers this use-case and more https://backstage.io/docs/plugins/internationalization

**Review by Rugvip — CHANGES_REQUESTED** [review]
Thank you! :tada:

Just a nit on the text customization. Wanna make sure we stick to our existing systems

File: plugins/kubernetes-react/config.d.ts
Comment (sebalaini) [reply plugins/kubernetes-react/config.d.ts:47]: besides this `buttonText` key is to give it a different text from the default one, Restart pod over Delete pod for example, wasn't meant to be used as a translation.

Comment (sebalaini) [reply plugins/kubernetes-react/config.d.ts:47]: OK, what if I remove the key in the documentation so that it's hidden and we merge this PR?
In this way, we can start to drop the k8 plugins internally in the company and then I'll work on adding the i18n API, or drop this key and use a single text for the button.

If in my company they are OK with using delete pod over restart pod what are your thoughts on this key? still useful or do we drop it?

**Comment by sebalaini** [comment]
hey @Rugvip @mclarke47 I committed a new change, let me know what you think about it :) 

I haven't implemented the i18n because I completely refactor the logic, instead of being able to use whatever text you want you can only set `true/false` and use **Restart pod** OR **Delete pod**

**Comment by sebalaini** [comment]
hey @Rugvip @mclarke47 @camilaibs can I get another review, please?

**Comment by mclarke47** [comment]
Need to wait on @Rugvip since he requested changes

File: docs/features/kubernetes/configuration.md
Comment (Rugvip) [inline docs/features/kubernetes/configuration.md:83]: Sorry but this is a bit of a step in the wrong direction as it reduced the configurability compared to before. Feedback remains the same though, the translation API needs to be used for text customization.

**Review by Rugvip — CHANGES_REQUESTED** [review]

File: docs/features/kubernetes/configuration.md
Comment (sebalaini) [reply docs/features/kubernetes/configuration.md:83]: OK I'll revert it and implement i18n

Comment (sebalaini) [reply docs/features/kubernetes/configuration.md]: done :) please give it an eye.

Comment (LegendSebastianoL) [reply docs/features/kubernetes/configuration.md]: hey @Rugvip can I get another review please 🙏

**Comment by LegendSebastianoL** [comment]
unstale, waiting approval

**Comment by mclarke47** [comment]
It seems we haven't used i18n for the button text as discussed?

**Comment by sebalaini** [comment]
> It seems we haven't used i18n for the button text as discussed?

can you point me to some examples?

I added `plugins/kubernetes-react/src/translation.ts` that is called inside `PodDeleteButton.tsx` 
I can translate only what I know, the key in the `app-config.yaml` how can it be translated?

this is the [commit](https://github.com/backstage/backstage/pull/24534/commits/4467e5b8f562856dac31bc2cf0444620ef47cdc4) to add i18n, is there something else I miss?

**Comment by LegendSebastianoL** [comment]
unstale

**Comment by LegendSebastianoL** [comment]
unstale.
Can I get clarification on what to do to fully implement the i18n 🙏 ?
I added something but looks like isn't enough.

**Comment by mclarke47** [comment]
@Rugvip is out this week, I'm sure he'll look when he gets a chance

**Comment by LegendSebastianoL** [comment]
unstale

**Comment by sebalaini** [comment]
can I please get a review? 🙏

If you are not interested anymore I'll see how to integrate it properly into our plugin version and still be able to update them with the upstream without wasting too much time every update.

File: plugins/kubernetes-react/src/components/Pods/PodDelete/PodDeleteButton.tsx
Comment (Rugvip) [inline plugins/kubernetes-react/src/components/Pods/PodDelete/PodDeleteButton.tsx:55]: Yep translation implementation looks good, but you also need to remove the other way of configuring the button text, this should be

```ts
const getButtonText = t('podDrawer.buttons.delete');
```

**Review by Rugvip — COMMENTED** [review]
Sorry realized there was a gap in the path to getting this ready, we need to get rid of the old button text config too

File: plugins/kubernetes-react/src/components/Pods/PodDelete/PodDeleteButton.tsx
Comment (LegendSebastianoL) [reply plugins/kubernetes-react/src/components/Pods/PodDelete/PodDeleteButton.tsx:55]: 🤔 so how can I change the `Delete Pod` text into `Restart pod`?

File: plugins/kubernetes-react/src/translation.ts
Comment (Rugvip) [inline plugins/kubernetes-react/src/translation.ts:21]:
```suggestion
  id: 'kubernetes-react',
```

We skip the plugin prefix on these

File: plugins/kubernetes-react/src/translation.ts
Comment (Rugvip) [inline plugins/kubernetes-react/src/translation.ts:20]: Ah, you'll want to export this via the `/alpha` entry point too, otherwise you don't be able to configure the message using the ref.

Comment (sebalaini) [reply plugins/kubernetes-react/src/translation.ts:20]: 🤔 how exactly? do you have examples?

Comment (sebalaini) [reply plugins/kubernetes-react/src/translation.ts:20]: I guess I'm missing an important key point here on how to customize the button text without using the new key I added in the `app-config`

Comment (Rugvip) [reply plugins/kubernetes-react/src/translation.ts:20]: Ideally https://backstage.io/docs/plugins/internationalization#for-an-application-developer-overwrite-plugin-messages explains it all, but just in case, you'd do something like this:

```ts
import { kubernetesReactTranslationRef } from '@backstage/plugin-kubernetes-react/alpha';

const app = createApp({
  // ...
  __experimentalTranslations: {
    resources: [
      createTranslationMessages({
        ref: kubernetesReactTranslationRef,
        messages: {
          "podDrawer.buttons.delete": 'Restart Pod'
        }
      })
    ]
  }
  // ...
});
```

Comment (Rugvip) [reply plugins/kubernetes-react/src/translation.ts:20]: For the `/alpha` export, see https://backstage.io/docs/tooling/cli/build-system#subpath-exports

**Comment by sebalaini** [comment]
OK, I think I've done everything :)

File: docs/features/kubernetes/configuration.md
Comment (Rugvip) [inline docs/features/kubernetes/configuration.md:22]: @mclarke47 it might be nicer to skip the `frontend` level here, so just `kubernetes.podDelete.enabled`. Or it could even be flipped around to something like `kubernetes.enabledActions.podDelete`. I don't mind too much either way though, there's not really an established pattern here so far.

**Review by Rugvip — APPROVED** [review]
Yes, thank you! 👍 

Looks good to me now, just and optional comment to consider before merge by @mclarke47. He's probably out next week though, just a heads up

File: docs/features/kubernetes/configuration.md
Comment (sebalaini) [reply docs/features/kubernetes/configuration.md:22]: the `frontend` bit was raised by him a while ago.
image

**Pull request merged**
